### DISCUSSION TITLE
Array inputs and clad::jacobian

### DISCUSSION THREAD

**Discussion by Towdo** [vgvassilev/clad#968] [Q&A]
Hey,

I'm wondering if it might be possible to have array inputs for functions passed to `clad::jacobian` in the future.
So instead of the following example, which is taken from the README:
```
void h(double a, double b, double output[]) {
    output[0] = a * a * a;
    output[1] = a * a * a + b * b * b;
    output[2] = 2 * (a + b);
}
```
I'd like to have:
```
void h(double x[2], double output[]) {
    double a = x[0];
    double b = x[1];
    output[0] = a * a * a;
    output[1] = a * a * a + b * b * b;
    output[2] = 2 * (a + b);
}
```

Currently, this does cause the following error:
`clang-18: /home/llvm/include/llvm/ADT/SmallVector.h:304: T& llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::operator[](size_type) [with T = const clang::ValueDecl*; <template-parameter-1-2> = void; reference = const clang::ValueDecl*&; size_type = long unsigned int]: Assertion "idx < size()" failed.`

Is this something that might be possible in the future, or is there something inherently stopping us from achieving this?

**Comment by parth-07** [comment]
Hi @Towdo 

> Is this something that might be possible in the future, or is there something inherently stopping us from achieving this?

This would be difficult to implement because the array size information is typically lost when arrays are passed to functions as pointers. We need the array size information at compile-time to properly map derivatives in the output Jacobian matrix. For a concrete example, please consider the below case:

```cpp
void fn(double *x, double u, double *output) { ... }
```

If `x` size is `2`, then `jacobianMatrix[0][2]` will be the derivative of `output[0]` w.r.t `u`. However, if `x` size is 3, then `jacobianMatrix[0][3]` will be the derivative of `output[0] w.r.t `u`.

Your current example should be easy to implement because the size information is preserved (`double x[2]`). However, this is not true for the general case (`double *x`).
